### PR TITLE
Rover: Add FS_OPTION for FS during hold, and RTL_END_BEHAVE for actions at end of RTLs

### PR DESCRIPTION
**Pull request by Hwurzburg**
Alternate to #12301....allows FS to occur while in HOLD mode if desired...

### CODE REVIEW

**Comment by Hwurzburg** [comment]
@rmackay9  Randy, perhaps this should be changed to a bitmask type param, and name changed to FS_OPTIONS....bit 1, enable FS in HOLD mode, bit 2 reserved for global FS entry only if >x meters from HOME as we discussed in DevCall, etc....if you think that is a good idea, I will change it  and push it...

**Comment by Hwurzburg** [comment]
Changed to a rover options bitmask to allow additional  independent features in the future...
first two: Enable FS to occur in HOLD mode, and add the ability to change to LOITER for boats at end of RTLs instead of circling

Tested in SITL....vehicle tests tomorrow...

**Comment by Pedals2Paddles** [comment]
I suggest calling the parameter `fs_options`.  That will make it jive with the same concept I'm working on for copter too. https://github.com/ArduPilot/ardupilot/pull/7432

**Comment by Hwurzburg** [comment]
Started out that way, but could be used for other options not related to FS, just like QuadPlane's Q_OPTIONS

**Comment by Pedals2Paddles** [comment]
Ah ok, good point.

File: APMrover2/Parameters.h [hunk 1]
```diff
@@ -370,6 +370,9 @@ class ParametersG2 {
     // mission behave
     AP_Int8 mis_done_behave;
 
+    // Hold FS behave
+    AP_Int8 r_options;
```
Comment (WickedShell) [inline APMrover2/Parameters.h]: This is a really small storage space for options, only gives you 7 options (easily) total. It's probably worth making this a Int32.

File: APMrover2/Parameters.cpp [hunk 2]
```diff
@@ -603,6 +603,14 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     // @Path: ../libraries/AC_Avoidance/AP_OAPathPlanner.cpp
     AP_SUBGROUPINFO(oa, "OA_", 45, ParametersG2, AP_OAPathPlanner),
 
+    // @Param: R_OPTIONS
+    // @DisplayName: Rover Options
+    // @Description: Bitmask to enable Rover options:
+    // @Values: 0:None,1:FS allowed in HOLD mode, 2:Enable loiter at end of RTLs
+    // @Bitmask: 0:Enable FS in HOLD mode, 1:Enable loiter at end of RTLs
+    // @User: Advanced
+    AP_GROUPINFO("R_OPTIONS", 46, ParametersG2, r_options, 0),
```
Comment (WickedShell) [inline APMrover2/Parameters.cpp]: This should probably not be prefixed with `R_`

**Comment by rmackay9** [comment]
As discussed on the dev call let's split this into two parameters, FS_OPTIONS and RTL_DONE_BEHAVE

File: APMrover2/Parameters.h (hunk 1 above)
Comment (Hwurzburg) [reply APMrover2/Parameters.h]: done in new commit

File: APMrover2/Parameters.cpp (hunk 2 above)
Comment (Hwurzburg) [reply APMrover2/Parameters.cpp]: changed per devcall

**Comment by Hwurzburg** [comment]
@rmackay9 changed per request and re-tested in SITL....this should be merged after #12358 to allow param numbering to be correct (although there will probably be a merge conflict anyway)

**Comment by Hwurzburg** [comment]
rebased, moved var declarations to end of G2

File: APMrover2/Rover.h [hunk 3]
```diff
@@ -490,6 +491,15 @@ class Rover : public AP_HAL::HAL::Callbacks {
 
     // Simple mode
     float simple_sin_yaw;
+
+    enum FS_Options {
```
Comment (rmackay9) [inline APMrover2/Rover.h]: Let's move this up below the "enum Failsafe_Action" and rename it to "Failsafe_Options" to be consistent.

File: APMrover2/Rover.h [hunk 4]
```diff
@@ -490,6 +491,15 @@ class Rover : public AP_HAL::HAL::Callbacks {
...
+        OPTION_HOLD_FS_ALLOW =  (1<<0)
+    };
+
+    enum Rtl_End_Behave {
```
Comment (rmackay9) [inline APMrover2/Rover.h]: let's move RTL_End_Behave up under cruise_learn_t and rename it to "RTL_Done_Behave" (note capital T and L) to be more consistent with the existing Mis_Done_Behave enum.  Perhaps just rename the options to "HOLD" and "LOITER" or maybe "RTL_DONE_BEHAVE_HOLD", etc.  sorry to be picky.

File: APMrover2/Rover.h (hunk 3 above)
Comment (Hwurzburg) [reply APMrover2/Rover.h]: done

File: APMrover2/Rover.h (hunk 4 above)
Comment (Hwurzburg) [reply APMrover2/Rover.h]: done....I think

**Comment by Hwurzburg** [comment]
will work on the new comments after dinner....

**Comment by Hwurzburg** [comment]
think I have all changes requested in now...

**Comment by rmackay9** [comment]
@Hwurzburg, I've created a cut-down version of this PR which includes only the FS_OPTIONS change.  https://github.com/ArduPilot/ardupilot/pull/12491.  If you're OK with that I'll merge it to master.

The reason I haven't merged the RTL_DONE_BEHAVE is that, although it works (I tested it in SITL) it works (I think) because of a race condition and in general we should not be setting the mode from within a mode.  Instead we should be implementing loiter or hold from within RTL.

I'm concerned that if we do this setting of mode from within RTL, we will paint ourselves into a corner because it starts getting complex when we consider RTL-within-Auto.  If you peek at how Auto executes an RTL command, it actually called the mode_rtl.update().  This means that theoretically if Auto ran an RTL command, then once the RTL completed it might execute the RTL_DONE_BEHAVE action and switch out of Auto and into Loiter.  But we already have a MIS_DONE_BEHAVE which controls what Auto should do when it completes.

Also we need to consider that the RTL may not be the last command in the mission (a user could reasonably set up a mission where the vehicle drives to a location, then comes home, then drives out to another location, etc).  In fact, this PR does work even in these complex situations but I think it's only because of a race condition.  I.e. auto executes the next command before RTL has a chance to switch the vehicle to Loiter mode.

So.. basically we need to call Loiter and Hold from within RTL instead of switching modes.  This is similar to the work we need to do to implement Hold from within Auto.

**Comment by Hwurzburg** [comment]
sounds good, but circling at full throttle in a tight circle isnt very good current behavior either....so I will start working on that next....so we need to change land rover behavior at RTL ends from mode HOLD to stop vehicle?

**Comment by rmackay9** [comment]
Ok, great, I'll have a look at the new PR.  Let's close this one.  the code stays so we can always revive this particular PR if necessary

**Pull request closed without merging**